core/vm/evm.go

```diff
@@ -277,23 +283,27 @@ func (evm *EVM) CallCode(caller ContractRef, addr common.Address, input []byte,
 	if !evm.Context.CanTransfer(evm.StateDB, caller.Address(), value) {
 		return nil, gas, ErrInsufficientBalance
 	}
-	var (
-		snapshot = evm.StateDB.Snapshot()
-		to       = AccountRef(caller.Address())
-	)
-	// Initialise a new contract and set the code that is to be used by the EVM.
-	// The contract is a scoped environment for this execution context only.
-	contract := NewContract(caller, to, value, gas)
-	contract.SetCallCode(&addr, evm.StateDB.GetCodeHash(addr), evm.StateDB.GetCode(addr))
-
-	ret, err = run(evm, contract, input, false)
+	var snapshot = evm.StateDB.Snapshot()
```

Comment:
If we execute a precompile we don't need to take a snapshot (and later revert),
because the precompiles are stateless right?

Reply:
Maybe we can change logic s.th. we only take a snapshot (and revert) with a normal delegatecall

Reply:
Ahh nvm saw your comment down below

Reply:
Well maybe not... the comment down below only talks about staticcall not precompiles

Reply:
We might get away with not taking a snapshot for a callcode to a precompile (also delegatecall, staticcall). However, there's no future-guarantee that all precompiles will never modify state. I didn't add that particular optimization, I'd rather save that for a future PR if it makes any difference (I think it doesn't really matter)